tweet_recommendations/other_methods/dbscan_based_method.py

```diff
@@ -0,0 +1,143 @@
+
+        self.clusterizer = DBSCAN(metric='manhattan', eps=epsilon, min_samples=MIN_SAMPLES, n_jobs=mp.cpu_count())
+        x["cluster_label"] = self.clusterizer.fit_predict(embeddings)
+
+        if self.verbose:
+            print("Clustering finished.")
+
+        x["hashtags"] = x["hashtags"].apply(lambda r: [elem["text"] for elem in r])
+
+        grouped_clusters = x.groupby(["cluster_label"])
+
+        if self.verbose:
+            print(f"Samples grouped into {len(np.unique(self.clusterizer.labels_))} clusters")
+            print("Building clusters' data")
+
+        self._centroids_data = []
+        self._corresponding_to_centroids_data_hashtags = []
+
+        for cluster_number, cluster in tqdm.tqdm(grouped_clusters, total=len(np.unique(x["cluster_label"])),
+                                                 disable=not self.verbose):
+            cluster_embeddings = np.vstack(cluster["embedding"].to_numpy())
+            cluster_center = np.mean(cluster_embeddings, axis=0).reshape(1, -1)
+            distances = distance.cdist(cluster_center, cluster_embeddings)[0]  # matrix of shape (1 x N) is returned
+
+            cluster_medoid_index = np.argmin(
+                distances)  # we are sure that it will not point to itself because mean center should be not existent
+
+            cluster_medoid = cluster_embeddings[cluster_medoid_index]
+            aggregated_hashtags = list(
+                chain.from_iterable(cluster["hashtags"].to_numpy()))
+            sorted_most_popular_hashtags = [hashtag for hashtag, count in
+                                            Counter(aggregated_hashtags).most_common()]
+
+            self._centroids_data.append(cluster_medoid)
+            self._corresponding_to_centroids_data_hashtags.append(sorted_most_popular_hashtags)
+
+        self._centroids_data = np.vstack(self._centroids_data)
+        self._corresponding_to_centroids_data_hashtags = np.asarray(self._corresponding_to_centroids_data_hashtags)
+
+        return self
+
+    def transform(self, x: Union[List[List[str]], List[str]]) -> np.ndarray:
+        """
+        For a given tweet/tweets embeddings recommend hashtags.
+        :param x: list of list of str or list of str. If first argument of x is a list is str, it is assumed that list
+            contains already lemmatized text. If single str is present as first element, it is assumed
+            that lemmatization has to be performed.
+        :return: Iterable of recommended hashtags.
+        """
+        embeddings = self.embedding_method.transform(x)
+        embeddings = embeddings if len(embeddings.shape) == 2 else embeddings.reshape(1, -1)
+        distances = cosine_distances(embeddings, self._centroids_data)
+        sorted_distances_indices = np.argsort(distances, axis=1)
+        recommended_hashtags = self._corresponding_to_centroids_data_hashtags[sorted_distances_indices]
+        result = self.post_process_result(recommended_hashtags)
+
+        return result
```

Comment:
zwracamy całą listę wszystkich hashtagów czy ograniczamy się do jakiegoś K?
przy ograniczeniu do K można trochę czasu zaoszczędzić na sortowaniu listy, choć nie wiem ile takich klastrów zwykle powstaje.

Reply:
Powstaje 40k, mimo wszystko sortowanie jest natychmiastowe

Reply:
A to ile jest hashtagów, że aż 40k klastrów? 😮

Reply:
45353 :wink: Dlatego parametr `neighbours_count` jest tak duży i pytanie czy nie będzie trzeba jeszcze większego. Zobaczę, jak się zachowuje przy manhattanie.

Reply:
OK, pewnie zwiększony filtr `minimal_hashtag_occurence` też może zmniejszyć liczbę tagów/klastrów